clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+  template <bool is_set> void get_set_attr(attribute attr, void *mem) {
+#define CASE(tag)                                                              \
+  case attribute::tag:                                                         \
+    if constexpr (is_set) {                                                    \
+      _##tag = *static_cast<decltype(_##tag) *>(mem);                          \
+    } else {                                                                   \
+      *static_cast<decltype(_##tag) *>(mem) = _##tag;                          \
+    }                                                                          \
+    break;
+    switch (attr) {
+      CASE(scale_type)
+      CASE(pointer_mode)
+      CASE(trans_a)
+      CASE(trans_b)
+    }
+#undef CASE
+  }
 
+  library_data_t _scale_type;
+  pointer_mode_t _pointer_mode = pointer_mode_t::host;
+  oneapi::mkl::transpose _trans_a = oneapi::mkl::transpose::nontrans;
+  oneapi::mkl::transpose _trans_b = oneapi::mkl::transpose::nontrans;
+
+  friend void matrix_transform(transform_desc_ptr transform_desc,
+                               const void *alpha, const void *a,
+                               matrix_layout_ptr a_desc, const void *beta,
+                               const void *b, matrix_layout_ptr b_desc, void *c,
+                               matrix_layout_ptr c_desc, queue_ptr q_ptr);
+};
+namespace detail {
+template <class T, class Talpha>
+void matrix_transform_impl(queue_ptr q_ptr, size_t a_rows, size_t a_cols,
+                           size_t a_ld, order_t a_order, T *a, size_t c_ld,
+                           order_t c_order, T *c, Talpha *alpha_data) {
+  if (a_order == order_t::col) {
+    q_ptr->submit([&](sycl::handler &cgh) {
+      cgh.parallel_for(sycl::range<2>(a_ld, a_cols), [=](sycl::id<2> index) {
+        if (index.get(0) < a_rows) {
+          size_t row_idx = index.get(0);
+          size_t col_idx = index.get(1);
+          size_t from_idx = a_ld * col_idx + row_idx;
+          size_t to_idx;
+          if (c_order == order_t::col) {
+            to_idx = c_ld * row_idx + col_idx;
+          } else {
+            to_idx = c_ld * col_idx + row_idx;
+          }
+          c[to_idx] = (*alpha_data) * a[from_idx];
+        }
+      });
+    });
+  } else {
+    q_ptr->submit([&](sycl::handler &cgh) {
+      cgh.parallel_for(sycl::range<2>(a_rows, a_ld), [=](sycl::id<2> index) {
+        if (index.get(1) < a_cols) {
+          size_t row_idx = index.get(0);
+          size_t col_idx = index.get(1);
+          size_t from_idx = a_ld * row_idx + col_idx;
+          size_t to_idx;
+          if (c_order == order_t::col) {
+            to_idx = c_ld * row_idx + col_idx;
+          } else {
+            to_idx = c_ld * col_idx + row_idx;
+          }
+          c[to_idx] = (*alpha_data) * a[from_idx];
+        }
+      });
+    });
+  }
+}
+} // namespace detail
+void matrix_transform(transform_desc_ptr transform_desc, const void *alpha,
```

Comment:
We do not need transpose matrix currently (at least for bitsandbytes).
For this function, we need impl below conversations:
col_major <=> row_major,
col_major <=> COL32,
col_major <=> COL4_4R2_8C,
col_major <=> COL32_2R_4R4

Reply:
done